zebra-consensus/src/transaction/check.rs

```diff
@@ -4,18 +4,49 @@
 
 use std::{borrow::Cow, collections::HashSet, convert::TryFrom, hash::Hash};
 
+use chrono::Utc;
+
 use zebra_chain::{
     amount::{Amount, NonNegative},
     block::Height,
     orchard::Flags,
     parameters::{Network, NetworkUpgrade},
     primitives::zcash_note_encryption,
     sapling::{Output, PerSpendAnchor, Spend},
-    transaction::Transaction,
+    transaction::{LockTime, Transaction},
 };
 
 use crate::error::TransactionError;
 
+/// Checks if the transaction's lock time allows this transaction to be included in a block.
+///
+/// Consensus rule:
+///
+/// > The transaction must be finalized: either its locktime must be in the past (or less
+/// > than or equal to the current block height), or all of its sequence numbers must be
+/// > 0xffffffff.
+///
+/// [`Transaction::lock_time`] validates the transparent input sequence numbers, returning [`None`]
+/// if they indicate that the transaction is finalized by them. Otherwise, this function validates
+/// if the lock time is in the past.
+pub fn lock_time_has_passed(tx: &Transaction, height: Height) -> Result<(), TransactionError> {
+    match tx.lock_time() {
+        Some(LockTime::Height(unlock_height)) => {
+            if height < unlock_height {
+                return Err(TransactionError::LockedUntilBlockHeight(unlock_height));
+            }
+        }
+        Some(LockTime::Time(unlock_time)) => {
+            if Utc::now() < unlock_time {
```

Comment:
Required Consensus Rule Change:

1. Consensus rules should not use the local node's time. Node clocks can be set to different times, which would cause a consensus split. (As an exception, we check the block time against the node's local time plus a 2 hour leeway for clock skew.)

Here is Zebra's block time:
https://github.com/ZcashFoundation/zebra/blob/43c6d013a7b120ccdcfcfa15c3de2e76b3c99853/zebra-chain/src/block/header.rs#L61-L63

2. The Bitcoin time docs say "greater than", not "greater than or equal to".

3. The Bitcoin height docs are different to what `zcashd` actually does. The block height/time must be less than the transaction lock height/time:
https://github.com/zcash/zcash/blob/1a7c2a3b04bcad6549be6d571bfdff8af9a2c814/src/main.cpp#L722

This means the Bitcoin doc is wrong for height, but correct for time:

> The transaction can be added to any block which has ~~this height or~~ higher.
> The transaction can be added to any block whose block time is greater than the locktime.

https://developer.bitcoin.org/devguide/transactions.html#locktime-and-sequence-number

Reply:
Done through the following commits (oldest first):

- [fb6efe7](https://github.com/jvff/zebra/commit/fb6efe720ccfd98962359248c2c480080de2db30).
- [833af3e](https://github.com/jvff/zebra/commit/833af3e1ae5b749b02dd7a68ca43037f6a031798).
- [1f900fb](https://github.com/jvff/zebra/commit/1f900fbe7341235ee7b245273301a402ff7c5ae4).
- [55e88e6](https://github.com/jvff/zebra/commit/55e88e640881cea446ae9b15629fd43fdff3f6bb).

Difference with all four commits together can be seen [here](https://github.com/jvff/zebra/compare/fb6efe720ccfd98962359248c2c480080de2db30%5E..55e88e640881cea446ae9b15629fd43fdff3f6bb).